### ISSUE TITLE
0.41 -> 0.42 Breaking change for RPC: CallStack (from HasCallStack)

### ISSUE THREAD

**Issue by brockelmore** [dapphub/dapptools#541]
One of these two things causes a breaking change for cheat-code heavy tests. I will try to get a minimally reproduced example. Here is full trace:
```
CallStack (from HasCallStack):
    error, called at src/EVM/Symbolic.hs:54:14 in hevm-0.42.0-1B7A8J2pBT6B3cFKLMhsva:EVM.Symbolic
```
Here are the changes that would affect it:
- Default to on chain values for coinbase, timestamp, difficulty, blocknumber when rpc is provided
- Perform tx initialization (gas payment, value transfer) in hevm exec, hevm symbolic and hevm dapp-test.

I imagine its the second. Here is the full command:
```
export ETH_RPC_URL="https://localhost:8545"

export DAPP_SRC=./contracts
export DAPP_SOLC_VERSION=0.5.15
export SOLC_FLAGS="--optimize --optimize-runs 50000"

dapp build

block=$(seth block latest)

export DAPP_TEST_TIMESTAMP=$(seth --field timestamp <<< "$block")
export DAPP_TEST_NUMBER=$(seth --field number <<< "$block")
export DAPP_TEST_ORIGIN="0x683A78bA1f6b25E29fbBC9Cd1BFA29A51520De84"
export DAPP_TEST_CHAINED=99
export DAPP_TEST_ADDRESS="0x683A78bA1f6b25E29fbBC9Cd1BFA29A51520De84"
printf 'Running test for address %s\n' "$DAPP_TEST_ADDRESS"
LANG=C.UTF-8 dapp test --rpc-url $ETH_RPC_URL -v --match test_kill
```

**Comment by MrChico** [comment]
Interesting. The error is from [forcing a symbolic value to be concrete](https://github.com/dapphub/dapptools/blob/master/src/hevm/src/EVM/Symbolic.hs#L54), but your test should only be dealing with concrete values anyway, so I'm confused why this would happen. A minimal example would really help

**Comment by brockelmore** [comment]
Found it. Here minimally reproducible:

```
// SPDX-License-Identifier: GPL-3
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

pragma solidity >=0.4.23;

contract DSTest {
    event eventListener          (address target, bool exact);
    event logs                   (bytes);
    event log_bytes32            (bytes32);
    event log_named              (string val);
    event log_named_address      (bytes32 key, address val);
    event log_named_bytes32      (bytes32 key, bytes32 val);
    event log_named_bool         (bytes32 key, bool val);
    event log_named_decimal_int  (bytes32 key, int val, uint decimals);
    event log_named_decimal_uint (bytes32 key, uint val, uint decimals);
    event log_named_int          (bytes32 key, int val);
    event log_named_uint         (bytes32 key, uint val);
    event log_named_string       (bytes32 key, string val);

    bool public IS_TEST;
    bool public failed;

    constructor() internal {
        IS_TEST = true;
    }

    function fail() internal {
        failed = true;
    }

    function expectEventsExact(address target) internal {
        emit eventListener(target, true);
    }

    modifier logs_gas() {
        uint startGas = gasleft();
        _;
        uint endGas = gasleft();
        emit log_named_uint("gas", startGas - endGas);
    }

    function assertTrue(bool condition) internal {
        if (!condition) {
            emit log_bytes32("Assertion failed");
            fail();
        }
    }

    function assertEq(address a, address b) internal {
        if (a != b) {
            emit log_bytes32("Error: Wrong `address' value");
            emit log_named_address("  Expected", b);
            emit log_named_address("    Actual", a);
            fail();
        }
    }

    function assertEq(address a, address b, string memory name) internal {
        if (a != b) {
            emit log_bytes32("Error: Wrong `address' value");
            emit log_named(name);
            emit log_named_address("  Expected", b);
            emit log_named_address("    Actual", a);
            fail();
        }
    }

    function assertEq32(bytes32 a, bytes32 b) internal {
        assertEq(a, b);
    }

    function assertEq(bytes32 a, bytes32 b) internal {
        if (a != b) {
            emit log_bytes32("Error: Wrong `bytes32' value");
            emit log_named_bytes32("  Expected", b);
            emit log_named_bytes32("    Actual", a);
            fail();
        }
    }

    function assertEq(bytes32 a, bytes32 b, string memory name) internal {
        if (a != b) {
            emit log_bytes32("Error: Wrong `bytes32' value");
            emit log_named(name);
            emit log_named_bytes32("  Expected", b);
            emit log_named_bytes32("    Actual", a);
            fail();
        }
    }
    function assertEq(bool a, bool b, string memory name) internal {
        if (a != b) {
            emit log_bytes32("Error: Wrong `bytes32' value");
            emit log_named(name);
            emit log_named_bool("  Expected", b);
            emit log_named_bool("    Actual", a);
            fail();
        }
    }

    function assertEqDecimal(int a, int b, uint decimals) internal {
        if (a != b) {
            emit log_bytes32("Error: Wrong fixed-point decimal");
            emit log_named_decimal_int("  Expected", b, decimals);
            emit log_named_decimal_int("    Actual", a, decimals);
            fail();
        }
    }

    function assertEqDecimal(uint a, uint b, uint decimals) internal {
        if (a != b) {
            emit log_bytes32("Error: Wrong fixed-point decimal");
            emit log_named_decimal_uint("  Expected", b, decimals);
            emit log_named_decimal_uint("    Actual", a, decimals);
            fail();
        }
    }

    function assertEq(int a, int b) internal {
        if (a != b) {
            emit log_bytes32("Error: Wrong `int' value");
            emit log_named_int("  Expected", b);
            emit log_named_int("    Actual", a);
            fail();
        }
    }

    function assertEq(uint a, uint b) internal {
        if (a != b) {
            emit log_bytes32("Error: Wrong `uint' value");
            emit log_named_uint("  Expected", b);
            emit log_named_uint("    Actual", a);
            fail();
        }
    }

    function assertEq(string memory a, string memory b) internal {
        if (keccak256(abi.encodePacked(a)) != keccak256(abi.encodePacked(b))) {
            emit log_bytes32("Error: Wrong `string' value");
            emit log_named_string("  Expected", b);
            emit log_named_string("    Actual", a);
            fail();
        }
    }

    function assertEq(int a, int b, string memory name) internal {
        if (a != b) {
            emit log_bytes32("Error: Wrong `int' value");
            emit log_named(name);
            emit log_named_int("  Expected", b);
            emit log_named_int("    Actual", a);
            fail();
        }
    }

    function assertEq(uint a, uint b, string memory name) internal {
        if (a != b) {
            emit log_bytes32("Error: Wrong `uint' value");
            emit log_named(name);
            emit log_named_uint("  Expected", b);
            emit log_named_uint("    Actual", a);
            fail();
        }
    }

    function assertEq(string memory a, string memory b, string memory name) internal {
        if (keccak256(abi.encodePacked(a)) != keccak256(abi.encodePacked(b))) {
            emit log_bytes32("Error: Wrong `string' value");
            emit log_named(name);
            emit log_named_string("  Expected", b);
            emit log_named_string("    Actual", a);
            fail();
        }
    }

    function assertEq0(bytes memory a, bytes memory b) internal {
        bool ok = true;

        if (a.length == b.length) {
            for (uint i = 0; i < a.length; i++) {
                if (a[i] != b[i]) {
                    ok = false;
                }
            }
        } else {
            ok = false;
        }

        if (!ok) {
            emit log_bytes32("Error: Wrong `bytes' value");
            emit log_named_bytes32("  Expected", "[cannot show `bytes' value]");
            emit log_named_bytes32("  Actual", "[cannot show `bytes' value]");
            fail();
        }
    }
}

interface Hevm {
    function warp(uint) external;
    function roll(uint) external;
    function store(address,bytes32,bytes32) external;
    function load(address,bytes32) external returns (bytes32);
}

contract HEVMBug is DSTest {

    // --- constants
    bytes20 constant CHEAT_CODE =
        bytes20(uint160(uint(keccak256('hevm cheat code'))));

    address yamV3 = address(0x0AaCfbeC6a24756c20D41914F2caba817C0d8521);

    // --- helpers
    Hevm hevm;
    address me;

    function setUp() public {
        hevm = Hevm(address(CHEAT_CODE));
        me = address(this);
    }

    function test_static_length() public {
      for (uint256 i = 0; i < 2; i++) {
        bytes32 prev = hevm.load(yamV3, 0);
      }
    }
}
```

It occurs when performing load in a for loop. Will test with other cheats as well

**Comment by brockelmore** [comment]
Tested with others (store, warp, roll) and they all worked fine

**Comment by MrChico** [comment]
Thanks for the example @brockelmore , I have a fix for this in #544 , which fixes your example and another breaking example I found. If you have the time, you are welcome to try this fix out and see if it solves your larger project. To install:
```
git clone git@github.com:dapphub/dapptools.git && cd dapptools && git checkout rpcsymfix
nix-env -iA dapp -f .
```

**Comment by brockelmore** [comment]
It works!
